src/gui/emojimodel.h

```diff
@@ -0,0 +1,139 @@
+        , isCustom(isCustom)
+    {
+    }
+    Emoji() = default;
+
+    friend QDataStream &operator<<(QDataStream &arch, const Emoji &object)
+    {
+        arch << object.unicode;
+        arch << object.shortname;
+        return arch;
+    }
+
+    friend QDataStream &operator>>(QDataStream &arch, Emoji &object)
+    {
+        arch >> object.unicode;
+        arch >> object.shortname;
+        object.isCustom = object.unicode.startsWith("image://");
+        return arch;
+    }
+
+    QString unicode;
+    QString shortname;
+    bool isCustom = false;
+
+    Q_GADGET
+    Q_PROPERTY(QString unicode MEMBER unicode)
+    Q_PROPERTY(QString shortname MEMBER shortname)
+    Q_PROPERTY(bool isCustom MEMBER isCustom)
+};
+
+class EmojiCategoriesModel : public QAbstractListModel
+{
+public:
+    QVariant data(const QModelIndex &index, int role) const override;
+    int rowCount(const QModelIndex &parent = QModelIndex()) const override;
+
+protected:
+    enum Roles {
+        EmojiRole = 0,
+        LabelRole
+    };
+
+    QHash<int, QByteArray> roleNames() const override;
+
+private:
+    struct Category
+    {
+        QString emoji;
+        QString label;
+    };
+    static const std::vector<Category> categories;
+};
+
+class EmojiModel : public QObject
+{
+    Q_OBJECT
+
+    Q_PROPERTY(QVariantList model READ model NOTIFY modelChanged)
+    Q_PROPERTY(QAbstractListModel *emojiCategoriesModel READ emojiCategoriesModel CONSTANT)
+
+    Q_PROPERTY(QVariantList history READ history NOTIFY historyChanged)
+
+    Q_PROPERTY(QVariantList people MEMBER people CONSTANT)
+    Q_PROPERTY(QVariantList nature MEMBER nature CONSTANT)
+    Q_PROPERTY(QVariantList food MEMBER food CONSTANT)
+    Q_PROPERTY(QVariantList activity MEMBER activity CONSTANT)
+    Q_PROPERTY(QVariantList travel MEMBER travel CONSTANT)
+    Q_PROPERTY(QVariantList objects MEMBER objects CONSTANT)
+    Q_PROPERTY(QVariantList symbols MEMBER symbols CONSTANT)
+    Q_PROPERTY(QVariantList flags MEMBER flags CONSTANT)
+
+public:
+    explicit EmojiModel(QObject *parent = nullptr)
+        : QObject(parent)
+        , _emojiCategoriesModel(new EmojiCategoriesModel)
+    {
+    }
+
+    Q_INVOKABLE QVariantList history() const;
+    Q_INVOKABLE void setCategory(const QString &category);
+    Q_INVOKABLE void emojiUsed(const QVariant &modelData);
+
+    QVariantList model() const;
+    QAbstractListModel *emojiCategoriesModel() const;
+
+signals:
+    void historyChanged();
+    void modelChanged();
+
+private:
+    static const QVariantList people;
+    static const QVariantList nature;
+    static const QVariantList food;
+    static const QVariantList activity;
+    static const QVariantList travel;
+    static const QVariantList objects;
+    static const QVariantList symbols;
+    static const QVariantList flags;
+
+    QSettings _settings;
+    QString _category = "history";
+
+    std::unique_ptr<EmojiCategoriesModel> _emojiCategoriesModel;
```

Comment:
not sure I get the question
what is the place where this could be a problem ?

Reply:
Exactly that method `EmojiModel::emojiCategoriesModel()`. It can then not return longer `EmojiCategoryModel*` but must return `EmojiCategoryModel const *`.

Reply:
so if `EmojiModel::emojiCategoriesModel()` should allow external users of the API to modify the model maybe it should not be a `const` method
we are trying to work around this `const` attribute by making the code more complex. does not look sane to me

Reply:
ok